### PR TITLE
 [zend-ldap] php 8.1 & 8.2 compatibility fixes

### PR DESCRIPTION
**Pull request by marcing**
**Zend_Ldap:** php 8.1 & 8.2 compatibility fixes.
Tested on openldap 2.5.x - some test assertions had to be adjusted.

Removed calls to `ldap_sort` while keeping sort features.

[ldap_sort](https://www.php.net/manual/en/function.ldap-sort.php) got deprecated in php 7.0.0, removed in php 8.0.0

ported from https://github.com/laminas/laminas-ldap/commit/35162e69f2685400470bd5d67764050bc521011e and latest version of https://github.com/laminas/laminas-ldap/tree/2.18.x

### CODE REVIEW

File: phpunit.xml.dist
Comment (partikus) [inline phpunit.xml.dist:15]: folks, what's the valid approach for defining consts? `phpunit.xml` or `tests/TestConfiguration.ci.php`?
I would move them into `TestConfiguration` php file to keep consistency and customization that can be done via copying `tests/TestConfiguration.ci.php` to `tests/TestConfiguration.php`.

File: tests/Zend/Ldap/.ci/basedn.ldif
Comment (partikus) [inline tests/Zend/Ldap/.ci/basedn.ldif:1]: please keep consistency between files and if possible add empty line at the end of file

File: tests/Zend/Ldap/Node/RootDseTest.php
Comment (partikus) [inline tests/Zend/Ldap/Node/RootDseTest.php:94]: why there is no more subentry?

**Review by partikus — DISMISSED** [review]

File: tests/Zend/Ldap/.ci/import.txt
Comment (glensc) [inline tests/Zend/Ldap/.ci/import.txt:1]: why in hidden dir ".ci"? A more appropriate is a subdirectory in "tests/resources".

File: tests/Zend/Ldap/BindTest.php
Comment (glensc) [inline tests/Zend/Ldap/BindTest.php:184]: these should probably be version checked:

```php
if (PHP_VERSION_ID >= 80100) ...
```

or do the string differences come from OpenLDAP library?

File: phpunit.xml.dist
Comment (marcing) [reply phpunit.xml.dist:15]: I guess we can move both TESTS_ZEND_LDAP_HOST & TESTS_ZEND_LDAP_PORT, or just the host. We will see what the host will be for LDAP.

File: phpunit.xml.dist
Comment (marcing) [reply phpunit.xml.dist:33]: It might be needed if we decide to set TESTS_ZEND_LDAP_USE_START_TLS to true.

File: tests/Zend/Ldap/.ci/basedn.ldif
Comment (marcing) [reply tests/Zend/Ldap/.ci/basedn.ldif:1]: both ldif files not do not an empty line at the end, do you mean we should add empty line to both?

File: tests/Zend/Ldap/.ci/import.txt
Comment (marcing) [reply tests/Zend/Ldap/.ci/import.txt:1]: @glensc it depends only on OpenLDAP version (I changed to newer message on the current version).
For the moment all the code is only php 7.4 compatible. After we manage to integrate it with github actions, we will fix it for php 8.

File: tests/Zend/Ldap/BindTest.php
Comment (glensc) [reply tests/Zend/Ldap/BindTest.php:184]: could then check for ldap library version:
- https://github.com/composer/composer/blob/bd6a5019b3bf5edf13640522796f54accaad789e/src/Composer/Repository/PlatformRepository.php
- https://github.com/composer/composer/blob/bd6a5019b3bf5edf13640522796f54accaad789e/src/Composer/Platform/Runtime.php#L91-L102

```
ldap

LDAP Support => enabled
Total Links => 0/unlimited
API Version => 3001
Vendor Name => OpenLDAP
Vendor Version => 20603
SASL Support => Enabled
```

File: tests/Zend/Ldap/.ci/basedn.ldif
Comment (falkenhawk) [reply tests/Zend/Ldap/.ci/basedn.ldif:1]: no let's not add empty lines here. those are special files.

File: phpunit.xml.dist
Comment (falkenhawk) [reply phpunit.xml.dist:33]: please modify `tests/TestConfiguration.ci.php` file instead to configure ldap for CI. Those should not be configured by default in phpunit.xml.dist as we must not expect that anyone who would like to contribute to zf1s and runs test locally will have to install and configure ldap on local dev environment.

For the purpose of local tests, you should copy `phpunit.xml.dist` to `phpunit.xml` and modify it there. The `phpunit.xml` is added to `.gitignore` for the reason.

File: phpunit.xml.dist
Comment (falkenhawk) [reply phpunit.xml.dist:15]: https://github.com/zf1s/zf1/pull/159#discussion_r1043251113

File: tests/Zend/Ldap/.ci/basedn.ldif
Comment (glensc) [reply tests/Zend/Ldap/.ci/basedn.ldif:1]: not empty line but line terminator

File: tests/Zend/Ldap/BindTest.php
Comment (glensc) [reply tests/Zend/Ldap/BindTest.php:184]: my proposal is to just use whatever version you extract as "vendor version" and use it as ">=" and leave the previous strings to the "else" branch.

but the final say is to maintainers, do they want to keep the "older strings" present as well.

simplest and cleanest is just go on and use new strings. 

you could guess by file timestamps (zend tests vs openldap releases) what version it might had been, probably some 10+ year old 2.4.x:
- https://www.openldap.org/software/download/OpenLDAP/openldap-release/

Comment (marcing) [reply tests/Zend/Ldap/BindTest.php:184]: I think that LDAP library version (Vendor Version => 20603) or in my case: 20447, does not correspond to OpenLDAP server version:  2.5.13 in my case. And the error messages depend on OpenLDAP server version. I tested it on 2.5 and 2.6 is the latest. It's likely that the old messages were present in version 2.4 of OpenLDAP server like you suggest.

Comment (glensc) [reply tests/Zend/Ldap/BindTest.php:184]: I'm sure most of the errors come from client library or php extension.

since there aren't much strings in ext/ldap/ldap.c:
- https://github.com/php/php-src/blob/php-8.2.0/ext/ldap/ldap.c

I am guessing libldap.

for example this particular one can't come from ldap server:
```diff
            $this->assertContains('Can\'t contact LDAP server', $zle->getMessage());
            $this->assertContains('Failed to connect to LDAP server:', $zle->getMessage());
```

maybe only few come from server, like `No object found for`.

Comment (marcing) [reply tests/Zend/Ldap/BindTest.php:184]: @glensc 
The error message comes from:
https://github.com/openldap/openldap/blob/5e2fa8a213b8a5d14a1d6aaf6852ad4188db2c3f/libraries/libldap/error.c
I could not find any version with "Failed to connect to LDAP server" in it on github, but new string passes on all php versions it seems.

**Review by marcing — COMMENTED** [review]
Changes implemented.

**Review by marcing — COMMENTED** [review]
Changes implemented.

File: .github/workflows/tests.yml
Comment (falkenhawk) [inline .github/workflows/tests.yml:97]: this is not needed since 1389 is default port number and ports binding on line 94 are hardcoded anyway

File: .github/workflows/tests.yml
Comment (falkenhawk) [inline .github/workflows/tests.yml:125]: I don't think all of this is useful to have it. You may check which php version is used and what extensions are enabled in each action run when you expand the "Setup PHP" section

File: .github/workflows/tests.yml
Comment (falkenhawk) [inline .github/workflows/tests.yml:147]: I'd move this under the step below, "Import LDAP Configuration" and possibly rename the step to "Setup LDAP"

File: .github/workflows/tests.yml
Comment (falkenhawk) [inline .github/workflows/tests.yml:164]: these changes should be reverted

File: packages/zend-ldap/library/Zend/Ldap.php
Comment (falkenhawk) [inline packages/zend-ldap/library/Zend/Ldap.php:1013]: is there any alternative to `ldap_sort` for PHP 7+?

File: tests/Zend/Ldap/ChangePasswordTest.php
Comment (falkenhawk) [inline tests/Zend/Ldap/ChangePasswordTest.php:101]: is this change necessary? `$bind` is not used anywhere.

File: tests/Zend/Ldap/ConnectTest.php
Comment (falkenhawk) [inline tests/Zend/Ldap/ConnectTest.php:49]: is this change needed?

File: tests/Zend/Ldap/OriginalOfflineTest.php
Comment (falkenhawk) [inline tests/Zend/Ldap/OriginalOfflineTest.php:63]: is this change necesary? `\2f` should be a forward-slash.

File: tests/Zend/Ldap/SearchTest.php
Comment (falkenhawk) [inline tests/Zend/Ldap/SearchTest.php:167]: this test should be marked as skipped, with a comment why, instead of returning true.

File: tests/resources/openldap/docker-compose.yml
Comment (falkenhawk) [inline tests/resources/openldap/docker-compose.yml:29]: volumes mapping seem to be messed up. Could those be cleaned up, please?

**Review by falkenhawk — COMMENTED** [review]
thanks a lot for making the ldap tests to run. That was hard.
Please adjust the PR title and description to actually reflect what this is all about.

File: .github/workflows/tests.yml
Comment (glensc) [inline .github/workflows/tests.yml:127]: why the caps change? should be separate PR perhaps?

File: .github/workflows/tests.yml
Comment (glensc) [inline .github/workflows/tests.yml:131]: why the caps change? should be separate PR perhaps?

File: .github/workflows/tests.yml
Comment (glensc) [inline .github/workflows/tests.yml:134]: why the caps change? should be separate PR perhaps?

File: .github/workflows/tests.yml
Comment (glensc) [inline .github/workflows/tests.yml:138]: why the caps change? should be separate PR perhaps?

File: .github/workflows/tests.yml
Comment (glensc) [inline .github/workflows/tests.yml:151]: is this block in yaml needed? you can just bind mount `/docker-entrypoint-initdb.d` to the container:

- https://hub.docker.com/r/bitnami/openldap/#:~:text=Initializing%20a%20new%20instance

Comment (glensc) [reply .github/workflows/tests.yml:151]: in either case. remove `-e` from echo. not needed

File: packages/zend-ldap/library/Zend/Ldap.php
Comment (glensc) [inline packages/zend-ldap/library/Zend/Ldap.php:1004]: extra trailing space should be removed

File: packages/zend-ldap/library/Zend/Ldap/Collection/Iterator/Default.php
Comment (glensc) [inline packages/zend-ldap/library/Zend/Ldap/Collection/Iterator/Default.php:322]: psr-2 formatting requires you put space between if and brance. this applies to all new code

File: packages/zend-ldap/library/Zend/Ldap/Collection/Iterator/Default.php
Comment (glensc) [inline packages/zend-ldap/library/Zend/Ldap/Collection/Iterator/Default.php:335]: make private/protected?

File: tests/Zend/Ldap/BindTest.php
Comment (glensc) [inline tests/Zend/Ldap/BindTest.php:176]: remove added trailing whitespace

File: tests/Zend/Ldap/BindTest.php
Comment (glensc) [reply tests/Zend/Ldap/BindTest.php:184]: seems maintainers did not voice their opinion here. so can stay as you committed.

File: tests/Zend/Ldap/ConnectTest.php
Comment (glensc) [inline tests/Zend/Ldap/ConnectTest.php:68]: code style (psr-2 afaik), always include braces (`{}`) in expressions

File: tests/Zend/Ldap/OriginalConnectTest.php
Comment (glensc) [inline tests/Zend/Ldap/OriginalConnectTest.php:69]: this can be simplified (imho more readable):

```php
$port = defined('TESTS_ZEND_LDAP_PORT') ? TESTS_ZEND_LDAP_PORT: 0;
```

File: tests/resources/openldap/README.md
Comment (glensc) [inline tests/resources/openldap/README.md:4]:
```suggestion
docker-compose exec openldap \
```

File: tests/resources/openldap/docker-compose.yml
Comment (glensc) [reply tests/resources/openldap/docker-compose.yml:29]: perhaps manual load of ldif files, utilize the `/docker-entrypoint-initdb.d` by the image:

- https://hub.docker.com/r/bitnami/openldap/#:~:text=Initializing%20a%20new%20instance

File: tests/resources/openldap/ldif-extras/acl-mod.ldif
Comment (glensc) [inline tests/resources/openldap/ldif-extras/acl-mod.ldif:17]: missing newline end of the .ldif files should be Imho added

File: tests/resources/openldap/ldif-extras/zend-acl-mod.ldif
Comment (glensc) [inline tests/resources/openldap/ldif-extras/zend-acl-mod.ldif:8]: remove the `-` only one change in this file?

File: tests/Zend/Ldap/OriginalOfflineTest.php
Comment (marcing) [reply tests/Zend/Ldap/OriginalOfflineTest.php:63]: @falkenhawk yes this one is necessary, they changed it the same way in laminas. That's just the expected result now.

File: tests/resources/openldap/docker-compose.yml
Comment (partikus) [reply tests/resources/openldap/docker-compose.yml]: Good point! I've unified the setup script but there are some limitations of `/docker-entrypoint-initdb.d` e.g. LDAP instance is stopped before running scripts from  `/docker-entrypoint-initdb.d`. More details can be found in `init.sh`

File: .github/workflows/tests.yml
Comment (glensc) [inline .github/workflows/tests.yml:150]: remove commented out block

File: tests/resources/openldap/docker-entrypoint-initdb.d/init.sh
Comment (glensc) [inline tests/resources/openldap/docker-entrypoint-initdb.d/init.sh:44]: more readable if you put each file on their own line.

```suggestion
files=(
    "manager.example.com.ldif"
    "test.example.com.ldif"
    "user1.example.com.ldif"
)
```

File: tests/resources/openldap/docker-compose.yml
Comment (glensc) [reply tests/resources/openldap/docker-compose.yml]: I've checked the script you shared, perhaps `LDAP_CUSTOM_SCHEMA_DIR` can be just used for the load?
- https://github.com/bitnami/containers/blob/fccaa4c4a4d7755c19c2e02ddef7ac3736dfcbb9/bitnami/openldap/2.6/debian-11/rootfs/opt/bitnami/scripts/libopenldap.sh#L361-L376

File: tests/TestConfiguration.ci.php
Comment (falkenhawk) [inline tests/TestConfiguration.ci.php:94]:
```suggestion
defined('TESTS_ZEND_LDAP_PORT') || define('TESTS_ZEND_LDAP_PORT', 1389);
```

File: tests/TestConfiguration.ci.php
Comment (falkenhawk) [inline tests/TestConfiguration.ci.php:100]:
```suggestion
defined('TESTS_ZEND_LDAP_BIND_REQUIRES_DN') || define('TESTS_ZEND_LDAP_BIND_REQUIRES_DN', true);
```

File: packages/zend-ldap/library/Zend/Ldap.php
Comment (falkenhawk) [reply packages/zend-ldap/library/Zend/Ldap.php:1012]: Maybe this could be ported https://github.com/laminas/laminas-ldap/commit/35162e69f2685400470bd5d67764050bc521011e

**Review by falkenhawk — COMMENTED** [review]
I don't think any changes around `TESTS_ZEND_LDAP_PORT` are necessary. imo they can be reverted. Wouldn't the ldap client connect to port 389 by default, whether it's specified or not?

File: packages/zend-ldap/library/Zend/Ldap.php
Comment (falkenhawk) [reply packages/zend-ldap/library/Zend/Ldap.php:1012]: #168

**Review by falkenhawk — APPROVED** [review]

**Pull request merged**
